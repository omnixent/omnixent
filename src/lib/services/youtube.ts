import { ServiceArgs } from '.';
import { normalizeServiceArgs } from '../utils';

const endpoint = 'https://clients1.google.com/complete/search';

export function formatURI(options: ServiceArgs): string {
  const { t, hl } = normalizeServiceArgs(options);
  return `${endpoint}?&client=youtube&gs_ri=youtube&q=${t}&hl=${hl}`;
}

export function extractBody(response: string) {
  const body = response.replace(/^window\.google\.ac\.h\(/, '').replace(/\)$/, '');

  return JSON.parse(body)
    ?.flat()
    ?.flat()
    ?.filter((el: any) => typeof el === 'string');
}
