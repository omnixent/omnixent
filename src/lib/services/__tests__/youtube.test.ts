import { formatURI, extractBody } from '../youtube';

test('Testing google formatURI', () => {
  expect(formatURI({ term: 'Node.js', language: 'it', country: 'us' })).toMatchInlineSnapshot(
    `"https://clients1.google.com/complete/search?&client=youtube&gs_ri=youtube&q=node.js&hl=us-IT"`,
  );
  expect(
    formatURI({ term: 'blue jeans brands', language: 'en', country: 'de' }),
  ).toMatchInlineSnapshot(
    `"https://clients1.google.com/complete/search?&client=youtube&gs_ri=youtube&q=blue%20jeans%20brands&hl=de-EN"`,
  );
});

test('Testing google extractBody', () => {
  expect(
    extractBody(
      `window.google.ac.h(["node.js",[["node.js download",0],["node.js",0],["node.js server-side javascript",0],["node.js tutorial",0],["node.js cos\u0027Ã¨",0],["node.js express",0],["node.js windows 7",0],["node.js server-side javascript virus",0],["node.js path",0],["node.js wikipedia",0],["node.js what is",0],["node.js require",0],["node.js docker",0],["node.js websocket",0]],{"k":1,"q":"ByZYgqmh8L85teUO48aZdL7g8xE"}])`,
    ),
  ).toMatchSnapshot();
  expect(
    extractBody(
      `window.google.ac.h(["blue jeans brands",[["blue jeans brands",0,[19]],["blue jeans brands from the 90s",0,[19]],["blue jeans brands list",0,[19]],["top blue jeans brands",0,[22,30,19]],["expensive blue jeans brands",0,[22,30,19]],["target blue jeans brands",0,[22,30,19]],["mens blue jeans brands",0,[22,30,19]],["western blue jeans brands",0,[22,30,19]],["famous blue jeans brands",0,[22,30,19]],["sargent blue jeans brands",0,[22,30,19]]],{"k":1,"q":"ByZYgqmh8L85teUO48aZdL7g8xE"}])`,
    ),
  ).toMatchSnapshot();
});
