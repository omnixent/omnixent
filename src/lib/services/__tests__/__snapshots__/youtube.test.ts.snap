// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing google extractBody 1`] = `
Array [
  "node.js",
  "node.js download",
  "node.js",
  "node.js server-side javascript",
  "node.js tutorial",
  "node.js cos'Ã¨",
  "node.js express",
  "node.js windows 7",
  "node.js server-side javascript virus",
  "node.js path",
  "node.js wikipedia",
  "node.js what is",
  "node.js require",
  "node.js docker",
  "node.js websocket",
]
`;

exports[`Testing google extractBody 2`] = `
Array [
  "blue jeans brands",
  "blue jeans brands",
  "blue jeans brands from the 90s",
  "blue jeans brands list",
  "top blue jeans brands",
  "expensive blue jeans brands",
  "target blue jeans brands",
  "mens blue jeans brands",
  "western blue jeans brands",
  "famous blue jeans brands",
  "sargent blue jeans brands",
]
`;
